libraries/botbuilder-dialogs-adaptive/src/actions/forEachPage.ts

```diff
@@ -43,130 +29,76 @@ export interface ForEachPageConfiguration extends DialogConfiguration {
-                        break;
-                }
-            }
-        }
+        return super.configure(config);
+    }
 
-        return this;
+    public async beginDialog(dc: DialogContext, options?: O): Promise<DialogTurnResult> {
+        return await this.nextPage(dc);
     }
 
-    public getDependencies(): Dialog[] {
-        return this.actions;
+    protected async onEndOfActions(dc: DialogContext, result?: any): Promise<DialogTurnResult> {
+        return await this.nextPage(dc);
     }
 
-    public async beginDialog(sequence: SequenceContext, options: ForEachPageOptions): Promise<DialogTurnResult> {
-        // Ensure planning context
-        if (!(sequence instanceof SequenceContext)) { throw new Error(`${this.id}: should only be used within an AdaptiveDialog.`) }
-        if (!this.list) { throw new Error(`${this.id}: no list expression specified.`) }
-        if (!this.pageSize) { throw new Error(`${this.id}: no pageSize expression specified.`) }
-
-        // Unpack options
-        let { list, offset, pageSize } = options;
-        const memory = sequence.state;
-        if (list == undefined) { list = this.list.evaluate(this.id, memory) }
-        if (pageSize == undefined) { pageSize = this.pageSize.evaluate(this.id, memory) }
-        if (offset == undefined) { offset = 0 }
-
-        // Get next page of items
-        const page = this.getPage(list, offset, pageSize);
-
-        // Update current plan
-        if (page.length > 0) {
-            sequence.state.setValue(this.valueProperty, page);
-            const changes: ActionChangeList = {
-                changeType: ActionChangeType.insertActions,
-                actions: []
-            };
-            this.actions.forEach((action) => changes.actions.push({ dialogStack: [], dialogId: action.id }));
-            if (page.length == pageSize) {
-                // Add a call back into forEachPage() at the end of repeated actions.
-                // - A new offset is passed in which causes the next page of results to be returned.
-                changes.actions.push({
-                    dialogStack: [],
-                    dialogId: this.id,
-                    options: {
-                        list: list,
-                        offset: offset + pageSize,
-                        pageSize: pageSize
-                    }
-                });
+    protected async onBreakLoop(dc: DialogContext, actionScopeResult: ActionScopeResult): Promise<DialogTurnResult> {
+        return await dc.endDialog();
+    }
+
+    protected async onContinueLoop(dc: DialogContext, actionScopeResult: ActionScopeResult): Promise<DialogTurnResult> {
+        return await this.nextPage(dc);
+    }
+
+    protected onComputeId(): string {
+        return `ForEachPage[${this.itemsProperty}]`;
+    }
+
+    private async nextPage(dc: DialogContext): Promise<DialogTurnResult> {
+        const itemsProperty = new ExpressionEngine().parse(this.itemsProperty);
```

Comment:
>new ExpressionEngine().parse(this.itemsProperty); [](start = 30, length = 49)

cache parsed expression

Reply:
> > new ExpressionEngine().parse(this.itemsProperty); [](start = 30, length = 49)
> 
> cache parsed expression

Cached parsed expression in a private property.